akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -2270,3 +2271,70 @@ private[akka] final class StatefulMapConcat[In, Out](val f: () => In => Iterable
   override def toString = "StatefulMapConcat"
 
 }
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] final case class AggregateWithBoundary[In, Agg, Out](
+    allocate: () => Agg,
+    aggregate: (Agg, In) => Boolean,
+    harvest: Agg => Out,
+    emitOnTimer: Option[(Agg => Boolean, FiniteDuration)])
+    extends GraphStage[FlowShape[In, Out]] {
+
+  emitOnTimer.foreach {
+    case (_, interval) => require(interval.gteq(1.milli), s"timer(${interval.toCoarsest}) must not be smaller than 1ms")
+  }
+
+  val in: Inlet[In] = Inlet[In](s"${this.getClass.getName}.in")
+  val out: Outlet[Out] = Outlet[Out](s"${this.getClass.getName}.out")
+  override val shape: FlowShape[In, Out] = FlowShape(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new TimerGraphStageLogic(shape) with InHandler with OutHandler {
+
+      private[this] var aggregated: Agg = null.asInstanceOf[Agg]
+
+      override def preStart(): Unit = {
+        emitOnTimer.foreach {
+          case (_, interval) => scheduleWithFixedDelay(s"${this.getClass.getSimpleName}Timer", interval, interval)
+        }
+      }
+
+      override protected def onTimer(timerKey: Any): Unit = {
+        emitOnTimer.foreach {
+          case (isReadyOnTimer, _) => if (aggregated != null && isReadyOnTimer(aggregated)) harvestAndEmit()
+        }
+      }
+
+      // at onPush, isAvailable(in)=true hasBeenPulled(in)=false, isAvailable(out) could be true or false due to timer triggered emit
+      override def onPush(): Unit = {
+        if (aggregated == null) aggregated = allocate()
+        if (aggregate(aggregated, grab(in))) harvestAndEmit()
+        // the decision to pull entirely depend on isAvailable(out)=true, regardless of result of aggregate
```

Comment:
@johanandren added some comments here, since this part is tricky and easy to miss